Source/WebKit/Platform/cocoa/XPCUtilities.mm

```diff
@@ -26,17 +26,52 @@
 #include "config.h"
 #include "XPCUtilities.h"
 
+#include <wtf/WTFProcess.h>
+
 namespace WebKit {
 
+static auto constexpr messageNameKey = "message-name";
+static auto constexpr exitProcessMessage = "exit";
+
 void terminateWithReason(xpc_connection_t connection, ReasonCode, const char*)
 {
     // This could use ReasonSPI.h, but currently does not as the SPI is blocked by the sandbox.
     // See https://bugs.webkit.org/show_bug.cgi?id=224499 rdar://76396241
     if (!connection)
         return;
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    xpc_connection_kill(connection, SIGKILL);
-ALLOW_DEPRECATED_DECLARATIONS_END
+
+    // First give the process a chance to exit cleanly by sending a XPC message to request termination.
+    // If no reply is received, try xpc_connection_kill.
+    auto completionBlock = makeBlockPtr([connection = OSObjectPtr(connection)](xpc_object_t event) mutable {
+        if (!event)
+            return;
+        xpc_type_t type = xpc_get_type(event);
+        RELEASE_LOG_ERROR(IPC, "Received XPC reply to termination request of type: %{public}s", xpc_type_get_name(type));
+        if (xpc_get_type(event) == XPC_TYPE_DICTIONARY)
+            return;
+        ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+        xpc_connection_kill(connection.get(), SIGKILL);
+        ALLOW_DEPRECATED_DECLARATIONS_END
+    });
+
+    auto exitMessage = adoptOSObject(xpc_dictionary_create(nullptr, nullptr, 0));
+    xpc_dictionary_set_string(exitMessage.get(), messageNameKey, exitProcessMessage);
+    xpc_connection_send_message_with_reply(connection, exitMessage.get(), dispatch_get_main_queue(), completionBlock.get());
+}
+
+void handleXPCExitMessage(xpc_object_t event)
+{
+    if (xpc_get_type(event) == XPC_TYPE_DICTIONARY) {
+        auto* messageName = xpc_dictionary_get_string(event, messageNameKey);
+        if (messageName && !strcmp(messageName, exitProcessMessage)) {
+            RELEASE_LOG_ERROR(IPC, "Received exit message, exiting now.");
```

Comment:
It feels hackish to rely on the (misbehaving) child process to exit() itself. I'm not against it as a temporary workaround until we have a proper solution but this is not a good long term solution.

Also, why are we waiting on a response to the child before trying to `xpc_connection_kill()`? Why can't we send the XPC message and then call `xpc_connection_kill()` right away (i.e. race them).

Finally, it is not clear to me what guarantees that the child process is not suspended and can actually process the incoming XPC.

Reply:
Yes, I agree. We should work towards getting a reliable way to terminate the WebKit processes that does not involve the processes exiting themselves.

I changed the code to avoid the reply back, but instead race the termination requests as you suggested.

When the termination is requested by the XPC connection termination watchdog, we are holding a background assertion. So in that case we should not be suspended. However, there is also a call to `terminateWithReason` in `ProcessLauncher::terminateXPCConnection`, and it is not immediately clear if we're holding an assertion then.

Reply:
> > and it is not immediately clear if we're holding an assertion then.
> 
> Seems like we'd want to address this then.

Yes, I think so too. I think I'll land this as-is now, and then follow-up with a fix for this shortly.